resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+# Dart 2.0 Upper and Lower bounds
+
+leafp@google.com
+
+This documents the currently implemented upper and lower bound computation,
+modified to account for explicit nullability and the accompanying type system
+changes (including the legacy types).  In the interest of backwards
+compatibility, it does not try to fix the various issues with the existing
+algorithm.  
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+## Helper predicates
+
+The **TOP** predicate is true for any type which is in the equivalence class of
+top types.
+
+- **TOP**(`T?`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`T*`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`dynamic`) is true
+- **TOP**(`void`) is true
+- **TOP**(`FutureOr<T>`) is **TOP**(T)
+- **TOP**(T) is false otherwise
+
+The **OBJECT** predicate is true for any type which is in the equivalence class
+of `Object`.
+
+
+- **OBJECT**(`Object`) is true
+- **OBJECT**(`dynamic`) is true
+- **OBJECT**(`void`) is true
+- **OBJECT**(`FutureOr<T>`) is **OBJECT**(T)
+- **OBJECT**(T) is false otherwise
+
+The **BOTTOM** predicate is true for things in the equivalence class of `Never`.
+
+- **BOTTOM**(`Never`) is true
+- **BOTTOM**(`X&T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`X extends T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`T`) is false otherwise
+
+The **NULL** predicate is true for things in the equivalence class of `Null`
+
+- **NULL**(`Null`) is true
+- **NULL**(`T?`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T*`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on top and `Object` types.
+
+- **MORETOP**(`void`, `T`) = true
```

Comment:
I don't think it matters.  We're imposing a total order on equivalent types, so we can choose any order we want.  We've generally chosen the order that simplifies the structure of the type, which is why I chose this order for `?` and `*`.  That is, we choose `LUB(dynamic, FutureOr<dynamic>)` to be `dynamic`, since it's a "simpler" type.  So similarly, this gives us that `LUB(dynamic, dynamic?) == dynamic`.  Either choice is valid.

Reply:
Right, there's no soundness issue. It's just a matter of whether the given total order will be subtle and surprising to developers, or they won't even notice it because it's all so natural. ;)

But `void` and `dynamic` are perhaps extra robust here, because any equational theory we might add would be expected to satisfy `void? == void` and `dynamic? == dynamic`; `void*` would still be protected rather well (by rules against using such a value), but `dynamic*` might mean "`Object`, with unchecked member access", so the order `dynamic > dynamic*` is right.